addons/website_sale_wallet/models/gift_card.py

```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 -*-
+
+import re
+from secrets import token_hex
+
+from odoo import models, fields, api
+
+
+class GiftCard(models.Model):
+    # inherited Fields
+    _name = "gift.card"
+    _description = "gift card model"
+    _inherit = ['website.multi.mixin']
+    _order = 'id desc'
+    _sql_constraints = [
+        ('unique_gift_card_code', 'UNIQUE(code)',
+         'The gift card code must be unique.')
+    ]
+
+    # helpers
+    @api.model
+    def _generate_code(self):
+        return re.sub(r'(.{4})(?!$)', r'\1-', token_hex(8))
+
+    # Relational fields
+    currency_id = fields.Many2one('res.currency',
+                                  string='Currency',
+                                  readonly=True,
+                                  default=lambda self: self.env.company.currency_id.id,
+                                  required=True)
+    sale_order_purchase_id = fields.Many2one("sale.order", readonly=True, default=None)
+    wallet_transaction_ids = fields.One2many('wallet.transaction', 'gift_card_id')
```

Comment:
a gift card, id converted to a wallet_id that will have a lot of wallet.Transaction.

Gift card should not have the direct link to wallet.transaction but only a link to the wallet that consumed this gift card

Reply:
A gift card is converted to a wallet transaction (after adding it to a website).
A wallet transaction is either a sell order that we did pay using a wallet or a gift card that has been added.
so I don't see why it should be linked to the wallet and not the transaction?

Reply:
Each user has a wallet for each website.
The wallet contains a list of wallet transactions.
Each wallet transaction is a debit or a credit.
Each wallet transaction consists of a gift card (when we add a gift card) or a sale order line (when we pay with a wallet on order)

Reply:
> A wallet transaction is either a sell order that we did pay using a wallet or a gift card that has been added.

So we are agree, a gift card, is linked to 1 wallet.transaction.... so it shoud be a m2o and not a o2m...

All what the customer want is to know what the gift card become... if you avec the wallet.transaction that credit the wallet, you have after the link on tx to the wallet if needed.

But ok, it can be better to have the wallet.tx and not the wallet... but as a m2o...